File: manacher.py
# BOJ 13275 가장 긴 팰린드롬 부분 문자열


def manacher(s):
    m = len(s)
    a = [0] * m
    r = p = -1
    for i in range(1, m):
        if r < i - 1 + a[i - 1]:
            r = i - 1 + a[i - 1]
            p = i - 1
        
        if r >= i:
            a[i] = min(r - i, a[2 * p - i])
        
        while True:
            a[i] += 1
            if (i - a[i] == -1) or (i + a[i] == m):
                a[i] -= 1
                break
            
            if s[i - a[i]] != s[i + a[i]]:
                a[i] -= 1
                break
    return max(a)


s = input()
ans = 2 * manacher(s) + 1
ss = '#'
for i in range(len(s)):
    ss += s[i] + '#'
ans = max(ans, manacher(ss))
print(ans)

File: manacher.cpp
// BOJ 13275 가장 긴 팰린드롬 부분 문자열

#include <bits/stdc++.h>
#define sz size()
#define bk back()
#define fi first
#define se second

using namespace std;
typedef long long ll;
typedef pair<int, int> pii;

vector<int> manacher(string &s) {
    int n = 2 * s.sz + 1;
    vector<char> w(n);
    for (int i = 0; i < n; i++) {
        if (i & 1)
            w[i] = s[i / 2];
        else
            w[i] = '#';
    }

    int r = -1;
    int c = -1;
    vector<int> ret(n);
    for (int i = 1; i < n; i++) {
        if (r < i - 1 + ret[i - 1]) {
            r = i - 1 + ret[i - 1];
            c = i - 1;
        }

        if (i <= r)
            ret[i] = min(r - i, ret[2 * c - i]);

        while (0 < i - ret[i] && i + ret[i] < n - 1 && w[i - ret[i] - 1] == w[i + ret[i] + 1])
            ret[i]++;
    }

    return ret;
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    cout.tie(0);

    string s;
    cin >> s;

    vector<int> v = manacher(s);

    cout << *max_element(v.begin(), v.end());
}
